`flash-attn` Installation Problem

```bash
jovyan@w: ~/tofu$ pip install flash-attn --no-build-isolation

Collecting flash-attn
  Using cached flash_attn-2.7.4.post1.tar.gz (6.0 MB)
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [25 lines of output]
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "/tmp/pip-install-2pd30da5/flash-attn_be676ed8407c4c34a6df1fdb2fe64203/setup.py", line 171, in <module>
          _, bare_metal_version = get_cuda_bare_metal_version(CUDA_HOME)
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/tmp/pip-install-2pd30da5/flash-attn_be676ed8407c4c34a6df1fdb2fe64203/setup.py", line 89, in get_cuda_bare_metal_version
          raw_output = subprocess.check_output([cuda_dir + "/bin/nvcc", "-V"], universal_newlines=True)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/opt/saturncloud/envs/saturn/lib/python3.12/subprocess.py", line 466, in check_output
          return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/opt/saturncloud/envs/saturn/lib/python3.12/subprocess.py", line 548, in run
          with Popen(*popenargs, **kwargs) as process:
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/opt/saturncloud/envs/saturn/lib/python3.12/subprocess.py", line 1026, in __init__
          self._execute_child(args, executable, preexec_fn, close_fds,
        File "/opt/saturncloud/envs/saturn/lib/python3.12/subprocess.py", line 1955, in _execute_child
          raise child_exception_type(errno_num, err_msg, err_filename)
      FileNotFoundError: [Errno 2] No such file or directory: '/usr/local/cuda/bin/nvcc'
      
      
      torch.__version__  = 2.4.0.post301
      
      
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```

```
[85/85] /opt/saturncloud/envs/tofu/bin/nvcc --generate-dependencies-with-compile --dependency-output /tmp/pip-install-1hwf9yeo/flash-attn_e995e135f14f4106a270be574aa70dcc/build/temp.linux-x86_64-cpython-310/csrc/flash_attn/src/flash_fwd_split_hdim64_fp16_sm80.o.d -I/tmp/pip-install-1hwf9yeo/flash-attn_e995e135f14f4106a270be574aa70dcc/csrc/flash_attn -I/tmp/pip-install-1hwf9yeo/flash-attn_e995e135f14f4106a270be574aa70dcc/csrc/flash_attn/src -I/tmp/pip-install-1hwf9yeo/flash-attn_e995e135f14f4106a270be574aa70dcc/csrc/cutlass/include -I/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/torch/include -I/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/torch/include/torch/csrc/api/include -I/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/torch/include/TH -I/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/torch/include/THC -I/opt/saturncloud/envs/tofu/include -I/opt/saturncloud/envs/tofu/include/python3.10 -c -c /tmp/pip-install-1hwf9yeo/flash-attn_e995e135f14f4106a270be574aa70dcc/csrc/flash_attn/src/flash_fwd_split_hdim64_fp16_sm80.cu -o /tmp/pip-install-1hwf9yeo/flash-attn_e995e135f14f4106a270be574aa70dcc/build/temp.linux-x86_64-cpython-310/csrc/flash_attn/src/flash_fwd_split_hdim64_fp16_sm80.o -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_BFLOAT16_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ --expt-relaxed-constexpr --compiler-options ''"'"'-fPIC'"'"'' -O3 -std=c++17 -U__CUDA_NO_HALF_OPERATORS__ -U__CUDA_NO_HALF_CONVERSIONS__ -U__CUDA_NO_HALF2_OPERATORS__ -U__CUDA_NO_BFLOAT16_CONVERSIONS__ --expt-relaxed-constexpr --expt-extended-lambda --use_fast_math -gencode arch=compute_80,code=sm_80 -gencode arch=compute_90,code=sm_90 --threads 2 -DTORCH_API_INCLUDE_EXTENSION_H '-DPYBIND11_COMPILER_TYPE="_gcc"' '-DPYBIND11_STDLIB="_libstdcpp"' '-DPYBIND11_BUILD_ABI="_cxxabi1011"' -DTORCH_EXTENSION_NAME=flash_attn_2_cuda -D_GLIBCXX_USE_CXX11_ABI=0
      ninja: build stopped: subcommand failed.
      Traceback (most recent call last):
        File "/tmp/pip-install-1hwf9yeo/flash-attn_e995e135f14f4106a270be574aa70dcc/setup.py", line 486, in run
          urllib.request.urlretrieve(wheel_url, wheel_filename)
        File "/opt/saturncloud/envs/tofu/lib/python3.10/urllib/request.py", line 241, in urlretrieve
          with contextlib.closing(urlopen(url, data)) as fp:
        File "/opt/saturncloud/envs/tofu/lib/python3.10/urllib/request.py", line 216, in urlopen
          return opener.open(url, data, timeout)
        File "/opt/saturncloud/envs/tofu/lib/python3.10/urllib/request.py", line 525, in open
          response = meth(req, response)
        File "/opt/saturncloud/envs/tofu/lib/python3.10/urllib/request.py", line 634, in http_response
          response = self.parent.error(
        File "/opt/saturncloud/envs/tofu/lib/python3.10/urllib/request.py", line 563, in error
          return self._call_chain(*args)
        File "/opt/saturncloud/envs/tofu/lib/python3.10/urllib/request.py", line 496, in _call_chain
          result = func(*args)
        File "/opt/saturncloud/envs/tofu/lib/python3.10/urllib/request.py", line 643, in http_error_default
          raise HTTPError(req.full_url, code, msg, hdrs, fp)
      urllib.error.HTTPError: HTTP Error 404: Not Found
      
      During handling of the above exception, another exception occurred:
      
      Traceback (most recent call last):
        File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/torch/utils/cpp_extension.py", line 2104, in _run_ninja_build
          subprocess.run(
        File "/opt/saturncloud/envs/tofu/lib/python3.10/subprocess.py", line 526, in run
          raise CalledProcessError(retcode, process.args,
      subprocess.CalledProcessError: Command '['ninja', '-v', '-j', '127']' returned non-zero exit status 1.
      
      The above exception was the direct cause of the following exception:
      
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "/tmp/pip-install-1hwf9yeo/flash-attn_e995e135f14f4106a270be574aa70dcc/setup.py", line 526, in <module>
          setup(
        File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/__init__.py", line 117, in setup
          return distutils.core.setup(**attrs)
        File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/_distutils/core.py", line 186, in setup
          return run_commands(dist)
        File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/_distutils/core.py", line 202, in run_commands
          dist.run_commands()
        File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/_distutils/dist.py", line 983, in run_commands
          self.run_command(cmd)
        File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/dist.py", line 999, in run_command
          super().run_command(command)
        File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/_distutils/dist.py", line 1002, in run_command
          cmd_obj.run()
        File "/tmp/pip-install-1hwf9yeo/flash-attn_e995e135f14f4106a270be574aa70dcc/setup.py", line 503, in run
          super().run()
        File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/command/bdist_wheel.py", line 379, in run
          self.run_command("build")
        File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/_distutils/cmd.py", line 339, in run_command
          self.distribution.run_command(command)
        File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/dist.py", line 999, in run_command
          super().run_command(command)
        File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/_distutils/dist.py", line 1002, in run_command
          cmd_obj.run()
        File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/_distutils/command/build.py", line 136, in run
          self.run_command(cmd_name)
        File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/_distutils/cmd.py", line 339, in run_command
          self.distribution.run_command(command)
        File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/dist.py", line 999, in run_command
          super().run_command(command)
        File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/_distutils/dist.py", line 1002, in run_command
          cmd_obj.run()
        File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/command/build_ext.py", line 99, in run
          _build_ext.run(self)
        File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/_distutils/command/build_ext.py", line 365, in run
          self.build_extensions()
        File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/torch/utils/cpp_extension.py", line 868, in build_extensions
          build_ext.build_extensions(self)
        File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/_distutils/command/build_ext.py", line 481, in build_extensions
          self._build_extensions_serial()
        File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/_distutils/command/build_ext.py", line 507, in _build_extensions_serial
          self.build_extension(ext)
        File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/command/build_ext.py", line 264, in build_extension
          _build_ext.build_extension(self, ext)
        File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/_distutils/command/build_ext.py", line 562, in build_extension
          objects = self.compiler.compile(
        File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/torch/utils/cpp_extension.py", line 681, in unix_wrap_ninja_compile
          _write_ninja_file_and_compile_objects(
        File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/torch/utils/cpp_extension.py", line 1784, in _write_ninja_file_and_compile_objects
          _run_ninja_build(
        File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/torch/utils/cpp_extension.py", line 2120, in _run_ninja_build
          raise RuntimeError(message) from e
      RuntimeError: Error compiling objects for extension
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for flash-attn
  Running setup.py clean for flash-attn
Failed to build flash-attn
ERROR: Failed to build installable wheels for some pyproject.toml based projects (flash-attn)
```

> Hey just want to make sure if the steps before that went successfully?
> 
> conda create -n tofu python=3.10
> conda activate tofu
> conda install pytorch pytorch-cuda=11.8 -c pytorch -c nvidia
> conda install -c "nvidia/label/cuda-11.8.0" cuda-toolkit
> pip install -r requirements.txt

@pratyushmaini Yes these all went through. Below is the latest error message when I wiped the environment and re-ran it. Is it a problem with perhaps newer version `flash-attn` that doesn't have backwards compatibility?

> ```
> [85/85] /opt/saturncloud/envs/tofu/bin/nvcc --generate-dependencies-with-compile --dependency-output /tmp/pip-install-1hwf9yeo/flash-attn_e995e135f14f4106a270be574aa70dcc/build/temp.linux-x86_64-cpython-310/csrc/flash_attn/src/flash_fwd_split_hdim64_fp16_sm80.o.d -I/tmp/pip-install-1hwf9yeo/flash-attn_e995e135f14f4106a270be574aa70dcc/csrc/flash_attn -I/tmp/pip-install-1hwf9yeo/flash-attn_e995e135f14f4106a270be574aa70dcc/csrc/flash_attn/src -I/tmp/pip-install-1hwf9yeo/flash-attn_e995e135f14f4106a270be574aa70dcc/csrc/cutlass/include -I/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/torch/include -I/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/torch/include/torch/csrc/api/include -I/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/torch/include/TH -I/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/torch/include/THC -I/opt/saturncloud/envs/tofu/include -I/opt/saturncloud/envs/tofu/include/python3.10 -c -c /tmp/pip-install-1hwf9yeo/flash-attn_e995e135f14f4106a270be574aa70dcc/csrc/flash_attn/src/flash_fwd_split_hdim64_fp16_sm80.cu -o /tmp/pip-install-1hwf9yeo/flash-attn_e995e135f14f4106a270be574aa70dcc/build/temp.linux-x86_64-cpython-310/csrc/flash_attn/src/flash_fwd_split_hdim64_fp16_sm80.o -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_BFLOAT16_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ --expt-relaxed-constexpr --compiler-options ''"'"'-fPIC'"'"'' -O3 -std=c++17 -U__CUDA_NO_HALF_OPERATORS__ -U__CUDA_NO_HALF_CONVERSIONS__ -U__CUDA_NO_HALF2_OPERATORS__ -U__CUDA_NO_BFLOAT16_CONVERSIONS__ --expt-relaxed-constexpr --expt-extended-lambda --use_fast_math -gencode arch=compute_80,code=sm_80 -gencode arch=compute_90,code=sm_90 --threads 2 -DTORCH_API_INCLUDE_EXTENSION_H '-DPYBIND11_COMPILER_TYPE="_gcc"' '-DPYBIND11_STDLIB="_libstdcpp"' '-DPYBIND11_BUILD_ABI="_cxxabi1011"' -DTORCH_EXTENSION_NAME=flash_attn_2_cuda -D_GLIBCXX_USE_CXX11_ABI=0
>       ninja: build stopped: subcommand failed.
>       Traceback (most recent call last):
>         File "/tmp/pip-install-1hwf9yeo/flash-attn_e995e135f14f4106a270be574aa70dcc/setup.py", line 486, in run
>           urllib.request.urlretrieve(wheel_url, wheel_filename)
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/urllib/request.py", line 241, in urlretrieve
>           with contextlib.closing(urlopen(url, data)) as fp:
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/urllib/request.py", line 216, in urlopen
>           return opener.open(url, data, timeout)
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/urllib/request.py", line 525, in open
>           response = meth(req, response)
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/urllib/request.py", line 634, in http_response
>           response = self.parent.error(
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/urllib/request.py", line 563, in error
>           return self._call_chain(*args)
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/urllib/request.py", line 496, in _call_chain
>           result = func(*args)
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/urllib/request.py", line 643, in http_error_default
>           raise HTTPError(req.full_url, code, msg, hdrs, fp)
>       urllib.error.HTTPError: HTTP Error 404: Not Found
>       
>       During handling of the above exception, another exception occurred:
>       
>       Traceback (most recent call last):
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/torch/utils/cpp_extension.py", line 2104, in _run_ninja_build
>           subprocess.run(
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/subprocess.py", line 526, in run
>           raise CalledProcessError(retcode, process.args,
>       subprocess.CalledProcessError: Command '['ninja', '-v', '-j', '127']' returned non-zero exit status 1.
>       
>       The above exception was the direct cause of the following exception:
>       
>       Traceback (most recent call last):
>         File "<string>", line 2, in <module>
>         File "<pip-setuptools-caller>", line 34, in <module>
>         File "/tmp/pip-install-1hwf9yeo/flash-attn_e995e135f14f4106a270be574aa70dcc/setup.py", line 526, in <module>
>           setup(
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/__init__.py", line 117, in setup
>           return distutils.core.setup(**attrs)
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/_distutils/core.py", line 186, in setup
>           return run_commands(dist)
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/_distutils/core.py", line 202, in run_commands
>           dist.run_commands()
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/_distutils/dist.py", line 983, in run_commands
>           self.run_command(cmd)
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/dist.py", line 999, in run_command
>           super().run_command(command)
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/_distutils/dist.py", line 1002, in run_command
>           cmd_obj.run()
>         File "/tmp/pip-install-1hwf9yeo/flash-attn_e995e135f14f4106a270be574aa70dcc/setup.py", line 503, in run
>           super().run()
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/command/bdist_wheel.py", line 379, in run
>           self.run_command("build")
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/_distutils/cmd.py", line 339, in run_command
>           self.distribution.run_command(command)
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/dist.py", line 999, in run_command
>           super().run_command(command)
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/_distutils/dist.py", line 1002, in run_command
>           cmd_obj.run()
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/_distutils/command/build.py", line 136, in run
>           self.run_command(cmd_name)
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/_distutils/cmd.py", line 339, in run_command
>           self.distribution.run_command(command)
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/dist.py", line 999, in run_command
>           super().run_command(command)
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/_distutils/dist.py", line 1002, in run_command
>           cmd_obj.run()
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/command/build_ext.py", line 99, in run
>           _build_ext.run(self)
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/_distutils/command/build_ext.py", line 365, in run
>           self.build_extensions()
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/torch/utils/cpp_extension.py", line 868, in build_extensions
>           build_ext.build_extensions(self)
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/_distutils/command/build_ext.py", line 481, in build_extensions
>           self._build_extensions_serial()
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/_distutils/command/build_ext.py", line 507, in _build_extensions_serial
>           self.build_extension(ext)
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/command/build_ext.py", line 264, in build_extension
>           _build_ext.build_extension(self, ext)
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/setuptools/_distutils/command/build_ext.py", line 562, in build_extension
>           objects = self.compiler.compile(
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/torch/utils/cpp_extension.py", line 681, in unix_wrap_ninja_compile
>           _write_ninja_file_and_compile_objects(
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/torch/utils/cpp_extension.py", line 1784, in _write_ninja_file_and_compile_objects
>           _run_ninja_build(
>         File "/opt/saturncloud/envs/tofu/lib/python3.10/site-packages/torch/utils/cpp_extension.py", line 2120, in _run_ninja_build
>           raise RuntimeError(message) from e
>       RuntimeError: Error compiling objects for extension
>       [end of output]
>   
>   note: This error originates from a subprocess, and is likely not a problem with pip.
>   ERROR: Failed building wheel for flash-attn
>   Running setup.py clean for flash-attn
> Failed to build flash-attn
> ERROR: Failed to build installable wheels for some pyproject.toml based projects (flash-attn)
> ```

--- a/README.md
+++ b/README.md
@@ -23,8 +23,8 @@ The dataset is in QA format, making it ideal for use with popular chat models su
 ```
 conda create -n tofu python=3.10
 conda activate tofu
-conda install pytorch pytorch-cuda=11.8 -c pytorch -c nvidia
-conda install -c "nvidia/label/cuda-11.8.0" cuda-toolkit
+conda install pytorch pytorch-cuda=12.4 -c pytorch -c nvidia
+conda install -c "nvidia/label/cuda-12.4.0" cuda-toolkit
 pip install -r requirements.txt
 pip install flash-attn --no-build-isolation
 ```

